plugins/crypto/mbedtls/securitypolicy_mbedtls_common.c

```diff
@@ -289,6 +300,252 @@ mbedtls_decrypt_rsaOaep(mbedtls_pk_context *localPrivateKey,
+        return UA_STATUSCODE_BADOUTOFMEMORY;
+    outPrivateKey->length = len;
+    memcpy(outPrivateKey->data, c, outPrivateKey->length);
+
+    return UA_STATUSCODE_GOOD;
+}
+
+UA_StatusCode
+mbedtls_createSigningRequest(mbedtls_pk_context *localPrivateKey,
+                             mbedtls_pk_context *csrLocalPrivateKey,
+                             mbedtls_entropy_context *entropyContext,
+                             mbedtls_ctr_drbg_context *drbgContext,
+                             UA_SecurityPolicy *securityPolicy,
+                             const UA_String *subjectName,
+                             const UA_ByteString *nonce,
+                             UA_ByteString *csr,
+                             UA_ByteString *newPrivateKey) {
+    /* Check parameter */
+    if(!securityPolicy || !csr || !localPrivateKey || !csrLocalPrivateKey) {
+        return UA_STATUSCODE_BADINVALIDARGUMENT;
+    }
+
+    UA_StatusCode retval = UA_STATUSCODE_GOOD;
+    size_t ret = 0;
+    char *subj = NULL;
+
+    mbedtls_pk_free(csrLocalPrivateKey);
+
+    /* CSR has already been generated and private key only needs to be set
+     * if a new one has been generated. */
+    if(newPrivateKey && newPrivateKey->length > 0) {
+        mbedtls_pk_init(csrLocalPrivateKey);
+        mbedtls_pk_setup(csrLocalPrivateKey, mbedtls_pk_info_from_type(MBEDTLS_PK_RSA));
+
+        /* Set the private key */
+        if(UA_mbedTLS_LoadPrivateKey(newPrivateKey, csrLocalPrivateKey, entropyContext))
+            return UA_STATUSCODE_BADSECURITYCHECKSFAILED;
+
+        return UA_STATUSCODE_GOOD;
+    }
+
+    /* CSR is already created. Nothing to do */
+    if(csr && csr->length > 0)
+        return UA_STATUSCODE_GOOD;
+
+    /* Get X509 certificate */
+    mbedtls_x509_crt x509Cert;
+    mbedtls_x509_crt_init(&x509Cert);
+    UA_ByteString certificateStr = UA_mbedTLS_CopyDataFormatAware(&securityPolicy->localCertificate);
+    ret = mbedtls_x509_crt_parse(&x509Cert, certificateStr.data, certificateStr.length);
+    UA_ByteString_clear(&certificateStr);
+    if(ret)
+        return UA_STATUSCODE_BADCERTIFICATEINVALID;
+
+    mbedtls_x509write_csr request;
+    mbedtls_x509write_csr_init(&request);
+    /* Set message digest algorithms in CSR context */
+    mbedtls_x509write_csr_set_md_alg(&request, MBEDTLS_MD_SHA256);
+
+    /* Set key usage in CSR context */
+    if(mbedtls_x509write_csr_set_key_usage(&request, MBEDTLS_X509_KU_DIGITAL_SIGNATURE |
+                                                     MBEDTLS_X509_KU_DATA_ENCIPHERMENT |
+                                                     MBEDTLS_X509_KU_NON_REPUDIATION |
+                                                     MBEDTLS_X509_KU_KEY_ENCIPHERMENT) != 0) {
+        retval = UA_STATUSCODE_BADINTERNALERROR;
+        goto cleanup;
+    }
+
+    /* Add entropy */
+    if(entropyContext && nonce && nonce->length > 0) {
+        if(mbedtls_entropy_update_manual(entropyContext,
```

Comment:
Assign an additional result variable that is then checked in the if-condition.
Shorter lines, easier to read.